pkg/proxy/winuserspace/proxier.go

```diff
@@ -194,105 +178,136 @@ func (proxier *Proxier) SyncLoop() {
-// Pass proxyPort=0 to allocate a random port. The timeout only applies to UDP
-// connections, for now.
-func (proxier *Proxier) addServiceOnPort(service proxy.ServicePortName, protocol api.Protocol, proxyPort int, timeout time.Duration) (*serviceInfo, error) {
-	sock, err := newProxySocket(protocol, proxier.listenIP, proxyPort)
-	if err != nil {
-		return nil, err
-	}
-	_, portStr, err := net.SplitHostPort(sock.Addr().String())
-	if err != nil {
-		sock.Close()
-		return nil, err
+// addServicePortPortal starts listening for a new service, returning the serviceInfo.
+// The timeout only applies to UDP connections, for now.
+func (proxier *Proxier) addServicePortPortal(servicePortPortalName ServicePortPortalName, protocol api.Protocol, listenIP string, port int, timeout time.Duration) (*serviceInfo, error) {
+	// add the IP address.  Node port binds to all interfaces.
+	if listenIP != allAvailableInterfaces {
+		serviceIP := net.ParseIP(listenIP)
+		args := proxier.netshIpv4AddressAddArgs(serviceIP)
+		existed, err := proxier.netsh.EnsureIPAddress(args, serviceIP)
+		if err != nil {
+			return nil, err
+		}
+		if !existed {
+			glog.V(3).Infof("Added ip address to fowarder interface for service %q at %s:%d/%s", servicePortPortalName, listenIP, port, protocol)
+		}
 	}
-	portNum, err := strconv.Atoi(portStr)
+
+	// add the listener, proxy
+	sock, err := newProxySocket(protocol, net.ParseIP(listenIP), port)
 	if err != nil {
-		sock.Close()
 		return nil, err
 	}
 	si := &serviceInfo{
 		isAliveAtomic:       1,
-		proxyPort:           portNum,
+		portal:              portal{ip: listenIP, port: port, isExternal: false},
 		protocol:            protocol,
 		socket:              sock,
 		timeout:             timeout,
 		activeClients:       newClientCache(),
 		sessionAffinityType: api.ServiceAffinityNone, // default
-		stickyMaxAgeMinutes: 180,                     // TODO: parameterize this in the API.
 	}
-	proxier.setServiceInfo(service, si)
+	proxier.setServiceInfo(servicePortPortalName, si)
 
-	glog.V(2).Infof("Proxying for service %q on %s port %d", service, protocol, portNum)
-	go func(service proxy.ServicePortName, proxier *Proxier) {
+	glog.V(2).Infof("Proxying for service %q at %s:%d/%s", servicePortPortalName, listenIP, port, protocol)
+	go func(service ServicePortPortalName, proxier *Proxier) {
 		defer runtime.HandleCrash()
 		atomic.AddInt32(&proxier.numProxyLoops, 1)
 		sock.ProxyLoop(service, si, proxier)
 		atomic.AddInt32(&proxier.numProxyLoops, -1)
-	}(service, proxier)
+	}(servicePortPortalName, proxier)
 
 	return si, nil
 }
 
+func (proxier *Proxier) closeServicePortPortal(servicePortPortalName ServicePortPortalName, info *serviceInfo) error {
+	// turn off the proxy
+	err := proxier.stopProxy(servicePortPortalName, info)
+	if err != nil {
+		return err
+	}
+
+	// close the PortalProxy by deleting the service IP address
+	listenIP := info.portal.ip
```

Comment:
nit: this assignment feels unnecesary, just use `info.portal.ip` in both places.